qmetaobject/src/connections.rs

```diff
@@ -106,43 +214,100 @@ cpp_class!(
+    /// space, i.e. if the offset is less than 0 or greater or equal to
+    /// object's size. Object's size must be known at compile time.
+    pub fn from_offset<O: QObject + Sized>(offset: isize) -> Self {
+        assert!(
+            offset >= 0 && offset < std::mem::size_of::<O>() as isize,
+            "Signal is not part of the Object: offset {} is outside of type `{}` object's memory",
+            offset, std::any::type_name::<O>()
+        );
+        cpp!(unsafe [offset as "ptrdiff_t"] -> SignalCppRepresentation as "SignalCppRepresentation" {
+            return SignalCppRepresentation(offset);
+        })
+    }
+}
+
+/// High-level typed wrapper for a pointer to a C++/Qt signal.
+///
+/// While low-level `SignalCppRepresentation` operated on pointers to 'erased'
+/// member functions types, this struct adds type-safe behavior on top of that.
+///
+/// `Args` is a type that matches the argument of the signal.
 ///
 /// For example, a C++ signal with signature `void (MyType::*)(int, QString)` will be represented
-/// by the struct `CppSignal<fn(int, QString)>`
+/// by the `CppSignal<fn(int, QString)>` type.
 pub struct CppSignal<Args> {
     inner: SignalCppRepresentation,
     phantom: std::marker::PhantomData<Args>,
 }
 
 impl<Args> CppSignal<Args> {
+    /// Wrap low level type-erased `SignalCppRepresentation` and provide
+    /// high-level typed wrapper.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure that number, types and order of arguments strictly
+    /// matches between signal represented by `inner` and `Args`. Passing
+    /// incorrect information may result in Undefined Behavior.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate cpp;
+    /// # use qmetaobject::*;
+    /// fn object_name_changed() -> CppSignal<fn(QString)> {
+    ///     unsafe {
+    ///         CppSignal::new(cpp!([] -> SignalCppRepresentation as "SignalCppRepresentation"  {
+    ///             return &QObject::objectNameChanged;
+    ///         }))
+    ///     }
+    /// }
+    /// # fn main() {
+    /// #     let _ = object_name_changed();
+    /// # }
+    /// ```
     pub unsafe fn new(inner: SignalCppRepresentation) -> Self {
         CppSignal { inner, phantom: Default::default() }
     }
 }
 
+// see module-level docs
 impl<Args> Clone for CppSignal<Args> {
     fn clone(&self) -> Self {
         *self
     }
 }
 
+// see module-level docs
 impl<Args> Copy for CppSignal<Args> {}
 
-/// Types of signals constructed with the qt_signal! macro.
+/// Types of signals constructed with the `qt_signal!` macro.
 ///
-/// This type is empty, only its address within the corresponding object matters
+/// This type is empty, only its address within the corresponding object matters.
 ///
-/// Args represents the type of the arguments, similar to the CppSignal ones
+/// `Args` represents the type of the arguments, same as in `CppSignal`.
 pub struct RustSignal<Args> {
```

Comment:
True. The last thing I didn't get is how offset within struct converts into a correct signal id. I guess I should investigate qobject_impl.rs more closely. I tried `cargo expand` on derived types, but there are no constants, no expressions, no explanations — just magic numbers like `if c == 10u32` (tl;rd i'm too stupid for that yet).

Also I thought of opening an issue (and someday implementing it) that one of signal kinds stuff should be bound to the object type, because it's an offset into objects' concrete type after all. Then usage of `connect()` should become safer.

And one more thing. In C++ Qt convert signal arguments into slot arguments' types if there are (implicit?) conversion constructors in classes. Shouldn't we require Slot's `Args` to be `From/Into` signal args? Not sure it would work.

Reply:
Moved to tracking issue #100. Such high-level architecture concerns are beyond the scope of this PR (despite the naming, it is mostly about documenting existing stuff), and so should not be blocking it, I guess.

Reply:
@ogoffart 

I've been getting it all wrong this whole time, didn't I?

The only thing Qt is interested in, when it comes to signals/slots, is their unique number, which identifies them among others within class hierarchy. So, things like `&QObject::objectNameChanged` in C++ or their Rust counterpart `CppSignalRepresentation` are only meaningful as far as they can be converted to some kind of magic integer (`u32`), a key in a meta table of a class which can be used to call, retrieve method's arguments count & types, return type etc. But instead of (or in addition to?) having such table, MOC (in `Generator::generateMetacall`) or its Rust counterpart (during build.rs phase, in `qobject_impl.rs`) crafts special function `static_metacall` which does a lot of dispatch in run-time.

And the sole two reasons for using offsets in `RustSignal` are that offsets are:
 - unique within Rust struct (as long as filed is non zero-sized, so minimal bool is used);
 - easy to compute, and just as easy to generate code which computes them at run-time.

Ain't that right?

Reply:
Yes, that sounds right. 
In the end, we need to assign a signal id for that signal. For each QMetaObject, signal id are like lost id or method id but signal id are lower. And the trick is to convert from whatever native language representation to that signal id.  In C++, there is a way to convert from a pointer to member function to to that id.   In the current implementation, there is a way to convert from the offset to the rust signal within the object to that id.

Reply:
Sweet.

Then of course there's totally no reason of having `MemberFunctionPtr` as such. I'll rewrite that soon.

Thank you!